Multi channel fix 2

KateMoreva opened the pull request:
OMEZarr reading/writing refactoring

constantinpape commented on src/test/java/projects/local/LocalAutophagosomesEMTest.java:
```diff
@@ -0,0 +1,48 @@
+import projects.BaseTest;
+
+@Slf4j
+public class LocalAutophagosomesEMTest extends BaseTest {
```
As discussed, this test can be removed since the `Autophagosome` project is deprecated.

constantinpape commented on src/test/java/projects/local/OpenLocalConstantinNoTables.java:
```diff
@@ -0,0 +1,48 @@
+public class OpenLocalConstantinNoTables extends BaseTest {
+    @Test
+    public void showYX() throws SpimDataException {
+        this.spimData = (SpimData) new SpimDataOpener().openSpimData("/Volumes/emcf/pape/jil", ImageDataFormat.BdvN5);
```
Do you already mark these tests so that they are not run by the CI? (Since they require local data)

constantinpape replied:
This needs to be done for all the "OpenLocal" tests.

KateMoreva replied:
Due to org.junit.Test maven will not run this tests. On your machine you would be able to run them (also with "Run tests in" is intelij Idea)

constantinpape commented on src/test/java/projects/remote/OpenRemoteAutophagosomesCLEM.java:
```diff
@@ -0,0 +1,52 @@
+
+@Slf4j
+public class OpenRemoteAutophagosomesCLEM extends BaseTest {
+    public OpenRemoteAutophagosomesCLEM() throws SpimDataException {
```
Should be removed because this project is deprecated, see above.

KateMoreva replied:
removed

constantinpape reviewed (commented):
Thanks a lot for working on this @KateMoreva!
Great to see all the tests being added ;). Two comments/questions:
- You can remove the autophagosome project tests (it's deprecated)
- Do you disable the tests for local projects in the CI somehow? If not we need to look into how to do this.

tischi commented on src/main/java/org/embl/mobie/io/ome/zarr/loaders/N5OMEZarrImageLoader.java:
```diff
@@ -343,17 +334,32 @@ private double[] getXYZScale( OmeZarrMultiscales.CoordinateTransformations coord
-                        scale[scalesFirstIndexBackward - 1],
-                        scale[scalesFirstIndexBackward - 2] };
+            if (zarrAxes.hasChannels()) {
+                if (zarrAxes.is3DWithChannels() || zarrAxes.is2D()) {
```
What does `is2D()` return?
I think this should include the case `is3DWithChannels()`, doesn't it?

constantinpape replied:
I also find this confusing. I would assume that `is3DWithChannels` applies to the cases where we have axes `CZYX` or `TCZYX`. And in that case the scale that you set does not make sense (because we should also set a scale value for `Z`).

KateMoreva replied:
fixed

tischi commented on src/main/java/org/embl/mobie/io/ome/zarr/loaders/N5OMEZarrImageLoader.java:
```diff
@@ -343,17 +334,32 @@ private double[] getXYZScale( OmeZarrMultiscales.CoordinateTransformations coord
-                        scale[scalesFirstIndexBackward - 1],
-                        1.0
-                };
+                if (zarrAxes.containsXYZCoordinates()) {
```
what does `containsXYZCoordinates()`mean?
does that mean `is3D()`?

constantinpape replied:
@KateMoreva could you have a look here? I agree with @tischi that `containsXYZCoordinates` is confusing and `is3D` would be more intuitive.

KateMoreva replied:
renamed

tischi commented on src/main/java/org/embl/mobie/io/ome/zarr/loaders/N5OMEZarrImageLoader.java:
```diff
@@ -385,19 +391,36 @@ private VoxelDimensions getVoxelDimensions3D( int setupId, int datasetId ) {
-            if ( scale != null && unit != null ) {
-                return new FinalVoxelDimensions(unit, scale);
+            if (scale != null && unit != null) {
+                if (zarrAxes.is2D() || zarrAxes.is3DWithChannels() ) {
```
same question as above.

tischi commented on src/main/java/org/embl/mobie/io/ome/zarr/loaders/N5OMEZarrImageLoader.java:
```diff
@@ -478,6 +501,7 @@ public CacheControl getCacheControl() {
         return cache;
     }
 
+    // TODO: remove
```
Can we resolve this TODO?

KateMoreva replied:
yes

tischi commented on src/main/java/org/embl/mobie/io/ome/zarr/loaders/N5OMEZarrImageLoader.java:
```diff
@@ -548,7 +575,7 @@ private double[][] readMipmapResolutions() throws IOException {
                 long[] dimensions = getDatasetAttributes(getPathName(setupId, level)).getDimensions();
                 mipmapResolutions[level] = new double[3];
-                if (dimensions.length < 3 && dimensionsOfLevel0.length < 3) {
+                if (multiscale.axes.is2D() || multiscale.axes.is3DWithChannels()) {
```
again :)

tischi commented on src/main/java/org/embl/mobie/io/ome/zarr/util/N5OMEZarrCacheArrayLoader.java:
```diff
@@ -64,61 +66,27 @@ public A loadArray(final long[] gridPosition) throws IOException {
 
         if (block == null) {
-            return (A) zarrArrayCreator.createEmptyArray(gridPosition);
+            return (A) zarrArrayCreator.createEmptyArray( gridPosition );
```
I think you could remove the spaces. I probably added them because it is my coding style.

tischi commented on src/main/java/org/embl/mobie/io/ome/zarr/util/ZarrAxes.java:
```diff
@@ -130,12 +132,36 @@ public boolean hasChannels() {
-        List<String> zarrAxisList = getAxesList();
-        return zarrAxisList.indexOf("t");
+        List<String> reverseAxesList = Lists.reverse(getAxesList());
+        return reverseAxesList.indexOf("t");
```
I think you could just:

`return Lists.reverse(getAxesList()).indexOf("t");`

tischi commented on src/main/java/org/embl/mobie/io/ome/zarr/util/ZarrAxes.java:
```diff
@@ -130,12 +132,36 @@ public boolean hasChannels() {
+    // spatial: 0,1,2 (x,y,z)
+    public Map< Integer, Integer > spatialToZarr()
+    {
+        // TODO implement for all cases
```
Are there cases missing?

tischi commented on src/test/java/spimdata/OmeZarrV4S3SpimDataTests.java:
```diff
@@ -1,15 +1,26 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
 
-public class OmeZarrV4S3SpimDataTests {
+public class OmeZarrV4S3SpimDataTests < N extends NumericType< N > & RealType< N > >
```
Should this class be removed and joined with the below `OmeZarrS3V4Opener`?

KateMoreva replied:
no

tischi commented on src/test/java/ui/OmeZarrS3V4Opener.java:
```diff
@@ -10,9 +14,10 @@
 public class OmeZarrS3V4Opener {
```
As far as I can see there are no tests yet in this class?
Like `assertEquals(...)`...
Is this on purpose or did you want to add some?

KateMoreva replied:
It is a UI test (only for visual check), assertEquals not needed

tischi reviewed (commented):
Thanks a lot for the great work!

I added a few minor comments.

My main question is whether we now already implemented tests for **all** the data in

```
https://s3.embl.de/i2k-2020/ngff-example-data/v0.4/...
```

? 

For all those the files in this bucket we needs `assertEquals(...)` test for the dimensions of all the axes. 

Or is this still pending?

KateMoreva commented on pom.xml:
```diff
@@ -9,7 +9,7 @@
     <groupId>org.embl.mobie</groupId>
     <artifactId>mobie-io</artifactId>
-    <version>1.2.12-SNAPSHOT</version>
+    <version>1.2.12-multi-channel-fix-SNAPSHOT</version>
```
fixed

KateMoreva commented:
> Thanks a lot for the great work!
> 
> I added a few minor comments.
> 
> My main question is whether we now already implemented tests for **all** the data in
> 
> ```
> https://s3.embl.de/i2k-2020/ngff-example-data/v0.4/...
> ```
> 
> ?
> 
> For all those the files in this bucket we needs `assertEquals(...)` test for the dimensions of all the axes.
> 
> Or is this still pending?

ngff package contains separate tests for each dataset (could be done in one file, but current way allows as to write special tests for some of them)

constantinpape commented:
> ngff package contains separate tests for each dataset (could be done in one file, but current way allows as to write special tests for some of them)

I think that's a good solution.

constantinpape commented:
Everything we need is implemented now, so I will merge.
